addons/l10n_it_edi_doi/tests/test_amounts_and_warnings.py

```diff
@@ -0,0 +1,555 @@
+            line_form.product_uom_qty = 2
+            line_form.price_unit = 3000
+            line_form.save()
+            self.assertEqual(
+                order_form.l10n_it_edi_doi_warning,
+                "Pay attention, the threshold of your Declaration of Intent test 2019-threshold 1000 of 1,000.00\xa0€ is exceeded by 5,000.00\xa0€, this document included.\n"
+                "Invoiced: 0.00\xa0€; Not Yet Invoiced: 6,000.00\xa0€"
+            )
+
+    def test_invoice(self):
+        """
+        Ensure the amounts and warnings are computed correctly in the following flow:
+        We create a single invoice and post it.
+        """
+        declaration = self.declaration_1000
+        declaration_tax = declaration.company_id.l10n_it_edi_doi_tax_id
+
+        self.assertRecordValues(declaration, [{
+            'invoiced': 0.0,
+            'not_yet_invoiced': 0.0,
+            'remaining': 1000.0,
+        }])
+
+        invoice = self.env['account.move'].create({
+            'move_type': 'out_invoice',
+            'company_id': self.company.id,
+            'partner_id': declaration.partner_id.id,
+            'invoice_date': declaration.start_date,
+            'l10n_it_edi_doi_id': declaration.id,
+            'invoice_line_ids': [
+                Command.create({
+                    'name': 'declaration line',
+                    'quantity': 1,
+                    'price_unit': 1000.0,  # == declaration.threshold
+                    'tax_ids': [Command.set(declaration_tax.ids)],
+                }),
+                Command.create({
+                    # The line should be ignored since it does not use the special tax
+                    'name': 'not a declaration line',
+                    'quantity': 1,
+                    'price_unit': 2000.0,  # > declaration.threshold; not counted
+                    'tax_ids': False,
+                }),
+            ],
+        })
+        # The amounts have not changed since the invoice has not been posted yet.
+        self.assertRecordValues(declaration, [{
+            'invoiced': 0.0,
+            'not_yet_invoiced': 0.0,
+            'remaining': 1000.0,
+        }])
+        # There is no warning since posting the invoice would not exceed the threshold.
+        # (only lines with the special tax are counted)
+        self.assertEqual(invoice.l10n_it_edi_doi_warning, "")
+
+        # Update the declaration part of the invoice to exceed the threshold
+        invoice.invoice_line_ids[0].price_unit = 2000  # > declaration.threshold
+        # The warning appears with the amount after posting the invoice.
+        self.assertEqual(
+            invoice.l10n_it_edi_doi_warning,
+            "Pay attention, the threshold of your Declaration of Intent test 2019-threshold 1000 of 1,000.00\xa0€ is exceeded by 1,000.00\xa0€, this document included.\n"
+            "Invoiced: 2,000.00\xa0€; Not Yet Invoiced: 0.00\xa0€"
+        )
```

Comment:
We want to know what would happen if we posted the invoice.
Without the warning the user would have to go to the declaration of intent,
see what is remaining, compute the right amount to not exceed it etc.
Otherwise we only see something went wrong after we posted already.

E.g. this way we could split a line into 2 (1 exempt under the declaration and 1 not)
and determine the qty / prices by trial and error.

Reply:
Then the comment is a bit misleading IMO.
It sounds like the warning appears "after posting the invoice" :innocent:

Reply:
I clarified the comment